File: apps/desktop/src/features/chat/lib/message-grouping.ts
import type { TimelineMessage } from "@/shared/lib/mock-data";

/** Default continuity window (Slack-like). */
export const MESSAGE_GROUP_WINDOW_MS = 10 * 60 * 1000;

/**
 * Author key for grouping: user messages share one key; agent messages group
 * by agent + session; system / tool_summary never group.
 */
export function messageAuthorKey(m: TimelineMessage): string | null {
  if (m.role === "system") return null;
  if (m.kind === "tool_summary") return null;
  if (m.role === "user") return "user";
  const agent = m.agent ?? "agent";
  const session = m.sessionId ?? "";
  return `agent:${agent}:${session}`;
}

/**
 * True when `curr` should hide avatar/header as a continuation of `prev`
 * (same author, within the time window).
 */
export function isMessageGroupContinuation(
  prev: TimelineMessage | undefined,
  curr: TimelineMessage,
  windowMs: number = MESSAGE_GROUP_WINDOW_MS,
): boolean {
  if (!prev) return false;
  const prevKey = messageAuthorKey(prev);
  const currKey = messageAuthorKey(curr);
  if (!prevKey || !currKey || prevKey !== currKey) return false;

  const prevMs = prev.createdAtMs;
  const currMs = curr.createdAtMs;
  // Without timestamps, still collapse consecutive same-author bubbles.
  if (
    prevMs == null ||
    currMs == null ||
    !Number.isFinite(prevMs) ||
    !Number.isFinite(currMs) ||
    prevMs <= 0 ||
    currMs <= 0
  ) {
    return true;
  }
  return Math.abs(currMs - prevMs) <= windowMs;
}

/** Local calendar day key `YYYY-MM-DD`, or null when timestamp missing. */
export function localDayKey(ms: number | undefined): string | null {
  if (ms == null || !Number.isFinite(ms) || ms <= 0) return null;
  const d = new Date(ms);
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

/** Human day divider label from a local day key or ms. */
export function formatDayDividerLabel(ms: number): string {
  const d = new Date(ms);
  const today = new Date();
  const yesterday = new Date();
  yesterday.setDate(today.getDate() - 1);

  const sameDay = (a: Date, b: Date) =>
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate();

  if (sameDay(d, today)) return "Today";
  if (sameDay(d, yesterday)) return "Yesterday";
  return d.toLocaleDateString(undefined, {
    weekday: "short",
    month: "short",
    day: "numeric",
    year:
      d.getFullYear() === today.getFullYear() ? undefined : "numeric",
  });
}

/**
 * Whether to insert a day divider before `curr` given the previous message.
 * Requires both messages to have valid createdAtMs and different local days.
 */
export function shouldShowDayDivider(
  prev: TimelineMessage | undefined,
  curr: TimelineMessage,
): boolean {
  const currKey = localDayKey(curr.createdAtMs);
  if (!currKey) return false;
  if (!prev) return true;
  const prevKey = localDayKey(prev.createdAtMs);
  if (!prevKey) return true;
  return prevKey !== currKey;
}

File: apps/mobile/lib/ui/features/social/lib/message_grouping.dart
import 'package:minos/domain/message_sender_ext.dart';
import 'package:minos/domain/social_message.dart';
import 'package:minos/src/rust/api/minos.dart';

/// Default continuity window (Slack-like; matches Desktop `MESSAGE_GROUP_WINDOW_MS`).
const int messageGroupWindowMs = 10 * 60 * 1000;

/// Author key for grouping consecutive collaboration rows.
///
/// - Users group by account id (multi-member groups must not collapse across people).
/// - Agents group by bot_id + optional session id.
/// - Recalled rows still keep their author key so neighboring non-recalled
///   messages from the same author can continue grouping around them only when
///   callers decide; [isMessageGroupContinuation] treats recalled as a break
///   so system-style recall chrome does not steal avatar slots awkwardly.
String? messageAuthorKey(SocialChatMessage message) {
  if (message.isRecalled) return null;
  if (message.senderType == SenderType.agent || message.sender.isBot) {
    final session = message.agentSessionIdFromMessageId ?? '';
    final agentId = message.sender.identityId;
    return 'agent:$agentId:$session';
  }
  return message.sender.groupingKey;
}

/// True when [curr] should hide avatar/header as a continuation of [prev].
bool isMessageGroupContinuation(
  SocialChatMessage? prev,
  SocialChatMessage curr, {
  int windowMs = messageGroupWindowMs,
}) {
  if (prev == null) return false;
  final prevKey = messageAuthorKey(prev);
  final currKey = messageAuthorKey(curr);
  if (prevKey == null || currKey == null || prevKey != currKey) {
    return false;
  }

  final prevMs = prev.createdAtMs;
  final currMs = curr.createdAtMs;
  // Without valid timestamps, still collapse consecutive same-author rows.
  if (prevMs <= 0 || currMs <= 0) {
    return true;
  }
  return (currMs - prevMs).abs() <= windowMs;
}

/// Local calendar day key `YYYY-MM-DD`, or null when timestamp missing/invalid.
String? localDayKey(int? ms) {
  if (ms == null || ms <= 0) return null;
  final d = DateTime.fromMillisecondsSinceEpoch(ms, isUtc: false);
  final y = d.year.toString().padLeft(4, '0');
  final m = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$y-$m-$day';
}

/// Whether to insert a day divider before [curr] given the previous message.
bool shouldShowDayDivider(SocialChatMessage? prev, SocialChatMessage curr) {
  final currKey = localDayKey(curr.createdAtMs);
  if (currKey == null) return false;
  if (prev == null) return true;
  final prevKey = localDayKey(prev.createdAtMs);
  if (prevKey == null) return true;
  return prevKey != currKey;
}

/// Human day divider label (localized Chinese, matches Mobile UI language).
String formatDayDividerLabel(int ms, {DateTime? now}) {
  final d = DateTime.fromMillisecondsSinceEpoch(ms, isUtc: false);
  final today = now ?? DateTime.now();
  final yesterday = DateTime(today.year, today.month, today.day - 1);

  if (_isSameDay(d, today)) return '今天';
  if (_isSameDay(d, yesterday)) return '昨天';
  if (d.year == today.year) {
    return '${d.month}月${d.day}日';
  }
  return '${d.year}年${d.month}月${d.day}日';
}

/// Compact clock for message headers (`HH:mm`).
String formatMessageClock(int ms) {
  if (ms <= 0) return '';
  final d = DateTime.fromMillisecondsSinceEpoch(ms, isUtc: false);
  final hh = d.hour.toString().padLeft(2, '0');
  final mm = d.minute.toString().padLeft(2, '0');
  return '$hh:$mm';
}

/// Full locale timestamp for long-press / accessibility.
String formatMessageFullTimestamp(int ms) {
  if (ms <= 0) return '';
  final d = DateTime.fromMillisecondsSinceEpoch(ms, isUtc: false);
  final y = d.year.toString().padLeft(4, '0');
  final m = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$y-$m-$day ${formatMessageClock(ms)}';
}

bool _isSameDay(DateTime lhs, DateTime rhs) {
  return lhs.year == rhs.year && lhs.month == rhs.month && lhs.day == rhs.day;
}
